### ISSUE TITLE
Proposal: ExperimentalAttribute

### ISSUE THREAD

**Issue by njy** [dotnet/corefx#42672]
## The current situation
Sometimes there's the need to create (for personal/internal use) and maybe even release (for public consumption) some code that is useful in and on itself, but it's just not ready to be set in stone regarding its shape, api surface area, namespaces and so on.
Maybe we are talking about an early stage pre-v1.0 release so the expectations about stability of code/design/api may be clear, but maybe they are in a new release that is adding new capabilities along the way.

## The need
There should be a clear way of warning developers that the specific feature they are using is marked as experimental, meaning that it can change shape, break or disappear at any time in the near future, until it will get eventually stabilized.

## Solution
The new ExperimentalAttribute can clearly signal all that has been said above, with the optional custom message already familiar from the ObsoleteAttribute for which this new attribute is kind of in the opposite direction, time-wise: "obsolete" meaning "this thing had its time, don't use it anymore" while "experimental" meaning "it's not yet ready for prime time").

```csharp
namespace System
{
	[AttributeUsage(AttributeTargets.All, AllowMultiple = false, Inherited = true)]
	public sealed class ExperimentalAttribute : Attribute
	{
		public ExperimentalAttribute()
		public ExperimentalAttribute(string message)
		public string Message { get; }
	}
}
```

## The output
It should be possible to configure what output it would produce (info/warning/error), either per project/solution or globally.
Probably it would also be a good idea to be able to configure that via an editorconfig setting like other rules already existing:

```editorconfig
# Experimental features
dotnet_use_code_with_experimental_attribute = false : error
```
The above example means "do not allow using code marked as experimental, and threat such usage as an error".

The default should probably be ` = true : warning` meaning experimental code can be used, but they should output a warning.

![exp-attr-output](https://user-images.githubusercontent.com/1010086/69388529-68892680-0cc9-11ea-94f1-ced235b73287.png)

## Other options considered
A way to warn developers about code that may be problematic to use is the ObsoleteAttribute: the problem is that it is kind of heavy handed as already said by others, and the wording "obsolete" would probably be confusing for something "new".

There's also the pre-release story inside Nuget, but there the scope is package-wide so it's not granular enough, not suitable for specific classes/interfaces/methods/props/etc being added to an existing codebase: as a thought exercise imagine having the "obsolete" concept but only expressable package-wide for an entire nuget packge, that wouldn't cut it.
Additionally it's something available only in nuget packages, and only usable as a filter while adding new ones.
Instead it should be possible to open an existing project, set a line in editorconfig (or the options panel or similar), compile, and see if anything experimental is being used.
Also, it should be usable in codebases that may not be published as nuget packages but referenced in other ways only inside the company or team, and it would still be useful to communicate with other team members about the potential breakage that may ensue by using some pieces of code.

**Comment by Clockwork-Muse** [comment]
>A way to warn developers about code that may be problematic to use is the ObsoleteAttribute: the problem is that it is kind of heavy handed as already said by others, and the wording "obsolete" would probably be confusing for something "new".

`ObsoleteAttribute` is there to allow developers to move _away_ from that particular API, so reusing it this way is definitely counter to intention.  It's primarily for the case where you want to enable consumers to take new versions of existing packages without them needing to (immediately) change any code.

>There's also the pre-release story inside Nuget, but there the scope is package-wide so it's not granular enough, not suitable for specific classes/interfaces/methods/props/etc being added to an existing codebase

... The problem is that it feels like _everything_ new in a pre-release package should be marked with any supposed `ExperimentalAttribute`; after all, until you publish the final package, the entire API is "not final".  There might be some argument to a "confidence" level of some sort, but that seems better handled at the documentation level.

Really, probably the better way to handle incoming APIs in a NuGet package would be to improve the tooling such that if you're including a pre-release package you get warnings on any new APIs.

> Additionally it's something available only in nuget packages, and only usable as a filter while adding new ones.

I'm generally a firm believer that even single-consumer libraries should generally be packaged up separately - not due to the forlorn hope they'll be reused, but rather to manage dependencies and separating changes.

> Also, it should be usable in codebases that may not be published as nuget packages but referenced in other ways only inside the company or team, and it would still be useful to communicate with other team members about the potential breakage that may ensue by using some pieces of code.

You can (and should) have private NuGet servers.  Pre-release packages would still solve this issue.

**Comment by njy** [comment]
> `ObsoleteAttribute` is there to allow developers to move _away_ from that particular API, so reusing it this way is definitely counter to intention. It's primarily for the case where you want to enable consumers to take new versions of existing packages without them needing to (immediately) change any code.

Yep, that was my point. I was just pointing out explicitly that was something already in the framework that can, in genral, "warn" developers about code usage.
For example I've used that in the past (only internally in my team) as a signal for other devs with a custom message along the lines of "EXPERIMENTAL: better not to use this yet" but, although it served the purpose effectively as a warning sign, it always smelled badly and when thinking at a framework level feature it is surely something not to consider seriously.

> There might be some argument to a "confidence" level of some sort, but that seems better handled at the documentation level.

On this I certainly agree.

> Really, probably the better way to handle incoming APIs in a NuGet package would be to improve the tooling such that if you're including a pre-release package you get warnings on any new APIs.

I thought about that, but here's one of the problems I see: if you reference a package marked as pre-release and there's no special attributes to identify "new stuff" the tooling can only warn about the usage of _everything_ in that package, and that would render this path basically useless.
Also, if with "on any new APIs" you mean a warning only for stuff that has been added, the question became how would the tooling select "just the new stuff"? I can't think about a sort of automatic diffing of the api surface area between package versions because it would be too complicated, brittle and unreliable and because it would be even hard to just determine what the "base" version for the diffing should be (the previous version? the last non pre-relase one?). Even more, new stuff not experimental would be marked as a warning where that may not be the case.

Instead, if code authors would simply mark experimental stuff as experimental just like they already do with the obsolete ones, all these problems would go away.

**Comment by njy** [comment]
Also, it would be interesting to hear from F# people ( @cartermp ? others?) who I just discovered have had an ExperimentalAttribute from some time to see how they are using it, if they already encountered some problems, possible design flaws, how the tooling around that is or should be, etc.

**Comment by Clockwork-Muse** [comment]
Base version would most likely have to be previous version (in semantic versioning) non pre-release (maaaaayyybe overrideable).  Diffing a reflection-gleaned tree should be relatively painless (and trivially cacheable).

> Even more, new stuff not experimental would be marked as a warning where that may not be the case.

That's kinda my point.  To me, everything new in a prerelease is experimental (subject to some confidence level).  You're using a prerelease package, all bets are off.

**Comment by cartermp** [comment]
@njy although we have that attribute defined in FSharp.Core, it's rarely used in practice. We don't use it for F# itself.

**Comment by njy** [comment]
> Base version would most likely have to be previous version (in semantic versioning) non pre-release (maaaaayyybe overrideable). Diffing a reflection-gleaned tree should be relatively painless (and trivially cacheable).

I have some doubts that this would work in a reasonably predictable/stable way and without hurting perf/user experience. Having said so, I certainly get your pov.

> To me, everything new in a prerelease is experimental (subject to some confidence level). You're using a prerelease package, all bets are off.

Just thinking out loud here, but wouldn't that include minor "new" things like new overrides of existing methods while missing things that are changed internally (impl) without a change in the method signature?

I mean, the first one (catching new overrides of existing methods) can even be an interesting concept, like suppose you where previously calling a method with a non-passed optional param, and now that same line of code would match a new override or vice versa.
It's something to think about.

Also, something like this could be in theory (if feasible/stable/performant as said above) be just a code analyzer.

Any thoughts on this?

**Comment by Clockwork-Muse** [comment]
> Just thinking out loud here, but wouldn't that include minor "new" things like new overrides of existing methods while missing things that are changed internally (impl) without a change in the method signature?

... Why would you ever want to care about the internal implementation changing (assuming the same tests pass)?  Anything that did this should explicitly include only publicly-visible API surfaces.

> I mean, the first one (catching new overrides of existing methods) can even be an interesting concept, like suppose you where previously calling a method with a non-passed optional param, and now that same line of code would match a new override or vice versa.

Overload resolution prefers methods with fewer parameters in the face of optional arguments, probably for exactly this reason, so adding a new method with optional arguments would just be "standard new".  
Most library consumers (and existing automated analyzers) are going to look at you funny if you add an overload that only differs in removing an optional parameter, unless you are also mark the original deprecated.  At which point normal deprecated rules apply.  If the behavior differs significantly, you would be tarred and feathered (a new method name would be appropriate, most likely).

> Also, something like this could be in theory (if feasible/stable/performant as said above) be just a code analyzer.

... I don't know if analyzers get access to package versions, but potentially yes.

**Comment by njy** [comment]
> Overload resolution prefers methods with fewer parameters in the face of optional arguments, probably for exactly this reason, so adding a new method with optional arguments would just be "standard new".

It was just an example: if that is in fact the precedence for overload resolution then just use the opposite example, like there was a 3 params method with the last one optional, now and there's also a new 2 params overload.
But again, that particular one was just an example, the first that popped up in my mind and not necessarily a real use case.
I'll try to come up with a better one.

**Comment by sharwell** [comment]
I like the idea of `[Preliminary]`. One of my older breaking changes documents used `<preliminary/>` in a documentation comment and explained it as follows (the attribute would work the same way but have a different form):

> This library may include types and/or members which are designated as preliminary features. The preliminary feature designation is indicated by a clear note at the top of the associated documentation page. In the library source code, preliminary features are indicated by including the `<preliminary/>` element to the XML documentation comment for the feature. While preliminary features are much more "flexible" during the release cycle of the library, certain rules do apply in order to ensure the stronger rules provided for stable features will not be violated by a change to a preliminary feature. The following list includes examples of these rules, but other rules may be imposed by basic logical constraints. The API elements referred to in the following list are assumed to be restricted to the publicly-exposed API of the library. The terms "member" and "members" refer to any method, property, or event.
>
> * A member may only refer to a preliminary type in its signature if it is also marked preliminary.
> * An interface may only include a preliminary member if it is also marked preliminary.
> * An interface may only extend a preliminary interface if it is also marked preliminary.
> * A class may only include a preliminary abstract member if either it is also marked preliminary, or all constructors for the class are marked internal. This restriction also applies to abstract classes which do not implement an abstract member declared in a base class.

**Comment by tannergooding** [comment]
I'm not sure how this differs from or improves upon what we are already doing today.

APIs don't go into CoreFX until officially reviewed/approved. APIs in CoreFX are considered "preview" until we RTM and ship them as stable on NuGet (this is designated by `-preview`, `-alpha`, `-beta`, etc monikers).

We sometimes ship out of band packages as `.Experimental` alongside a given release. These are not "stable" and subject to change in the future, etc.

**Comment by Clockwork-Muse** [comment]
@tannergooding - I believe the proposal is not necessarily that we'd use it here, but that it would be included here for library (and tool) writers to use it themselves.

Although my position is otherwise the same as yours, I think: "unfinal packages are subject to change at any time".

**Comment by sharwell** [comment]
Without analyzer support, it's extremely easy to accidentally use a preliminary API. It's hit Roslyn users a few times. Regardless of the manner in which preliminary APIs are identified, it needs to be possible to create an analyzer that knows how to flag uses of them.

**Comment by njy** [comment]
> I like the idea of `[Preliminary]`. One of my older breaking changes documents used `<preliminary/>` in a documentation comment and explained it as follows (the attribute would work the same way but have a different form)

Yep, that's the general concept. Regarding the name either Preliminary, Experimental or something else would be ok, I came up with experimental and later discovered it was what already used in F# so it's probably more uniform.

**Comment by njy** [comment]
> I'm not sure how this differs from or improves upon what we are already doing today.

This would be more granular, for example. I don't know if it's the right way, but I had this need in the past and thought about a way to have something available in the framework to deal with those situations, just like the [Obsolete] attribute that helped me a lot in the past.

> APIs don't go into CoreFX until officially reviewed/approved. APIs in CoreFX are considered "preview" until we RTM and ship them as stable on NuGet (this is designated by `-preview`, `-alpha`, `-beta`, etc monikers).

As correctly said by @Clockwork-Muse I was not necessarily proposing the use in CoreFX itself (maybe, maybe not, it's not up to me): I created the issue here because that is where it should be defined to be broadly usable, just like the [Obsolete] attribute.

> We sometimes ship out of band packages as `.Experimental` alongside a given release. These are not "stable" and subject to change in the future, etc.

Yep, I saw that, but that requires different packages, probably different namespaces (eg: https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Experimental/MLContextExtensions.cs ) that should be later changed everywhere it has been referenced, even if the final shape of the API would remain the same.

If you think about it, you would still have the ability to have separate assemblies/packages if that is needed, but with the added bonus of warning developers about experimental stuff also inside already existing packages without the need to have new namespaces + later refactoring, and to warn them in granular way only about specific functionalities/classes/methods/props/etc.

Also, I know it's not a _"just add an ExperimentalAttribute class and you're done"_ kind of effort because of tooling support, documentation, etc: having said so, effort wise it may not be such a huge undertaking to have some initial results. The first step may be to create the attribute and generate a warning/error if needed, and additional features/tooling around such attribute may be added in the future on top of it, if its usage would grow over time and the community appreciate it.

Does it make sense?

**Comment by njy** [comment]
@Clockwork-Muse I thought a little bit about the analyzer approach or the likes.
Here's what it should probably do:
1. look at all the pkg deps available (also, sub deps?)
2. filter out only the ones with a preview/alpha/beta monikers
3. contact the pkg source (nuget + others) and, for every one, get a list of all the previous versions (is this possible?) to get the newest previous stable version, or do the same via an hypothetical specialized call
4. if one is found for each dependency, analyze all the types inside each assembly and, for each one, see if there has been a change in respect to the shape of the api
5. for each change, track its usage inside the current proj/sln
6. obviously more edge cases, special situations and whatnot

In general it would be cool, but I suspect it would not be an easily obtainable feature, on top of being a path full of hard situations to go over.

Thoughts?

**Comment by Clockwork-Muse** [comment]
>1. look at all the pkg deps available (also, sub deps?)
>2. filter out only the ones with a preview/alpha/beta monikers
>3. contact the pkg source (nuget + others) and, for every one, get a list of all the previous versions (is this possible?) to get the newest previous stable version, or do the same via an hypothetical specialized call
>4. if one is found for each dependency, analyze all the types inside each assembly and, for each one, see if there has been a change in respect to the shape of the api
>5. for each change, track its usage inside the current proj/sln
>6. obviously more edge cases, special situations and whatnot

This is what I'm thinking, yes.

1. I haven't looked into it, so I don't know if analyzers get access to package/solution information (as opposed to data from a loaded and referenced dll) - most of the info from a quick look is about writing an analyzer that we include with our own package.  We don't care about internal dependencies - public transitive ones also become our dependencies, though, if we use them.
2. If you can get package info, this is trivial
3. [There's an explicit API call for that](https://docs.microsoft.com/en-us/nuget/api/search-query-service-resource), and you can even exclude pre-release packages.  
4. Yes.  Since we don't need to actually load/execute the code inside them, you'd probably only need to load [into a reflection-only context](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.reflectiononlyload?view=netcore-3.0#System_Reflection_Assembly_ReflectionOnlyLoad_System_String_).  You would diff the trees/compare all symbols, **and store/cache the difference only**.  Almost certainly this can be computed on pre-release package load, once.
5. The analyzer checks if any symbol used is in the cached diff tree, and marks it, and marks it somehow.

<hr />

> ... but with the added bonus of warning developers about experimental stuff also inside already existing packages ...

If you publish a full, non pre-release, package with an "experimental" sub-section, you will earn my ire.  
This sort of behavior makes _some small sense_ when distribution is difficult and often requires that consumers build source on different machines, like in traditional C++.  It makes absolutely no sense when you have dependency management tools like NuGet.  Make a separate package for just the new stuff, or make a pre-release package with experimental stuff added.

Note that namespaces in C# are not connected to the package name (which sometimes drives me nuts, having come from Java), so separate namespaces are not required (and in fact several packages in corefx make use of this).  You don't necessarily need separate packages (just separate _versions_), although doing so can help manage things.

**Comment by sharwell** [comment]
@njy The approach of looking at monikers does not address cases where you want to ship a stable package with one or more APIs still marked as preliminary. It also doesn't allow the granularity to be controlled (e.g. the use of a preliminary API would force the entire assembly to be preliminary). NuGet already provides direct validation for prerelease packages at this granularity, so I don't see the value in duplicating it elsewhere.

> We sometimes ship out of band packages as .Experimental alongside a given release. 

It is not possible to add preliminary members to existing APIs using this approach.

> Also, I know it's not a "just add an ExperimentalAttribute class and you're done" kind of effort because of tooling support, documentation, etc: having said so, effort wise it may not be such a huge undertaking to have some initial results.

The [BannedApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/tree/master/src/Microsoft.CodeAnalysis.BannedApiAnalyzers) package could be updated with new analyzers for this scenario. It already contains shared infrastructure for examining code to ensure disallowed APIs (for an extensible definition of disallowed) are not used.

It would not be difficult to expand the analysis to cover the rules for applying `[Preliminary]` to code, and update [PublicApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/tree/master/src/PublicApiAnalyzers) to clearly indicate that these APIs are not considered "shipped".

**Comment by njy** [comment]
> Note that namespaces in C# are not connected to the package name [...], so separate namespaces are not required

Of course it is not required: my observation was that if it has been decided to use a different namespace to better mark the experimental state of the thing it was because of a need, and that need may be better covered by the proposed attribute, that would also not require a later refactoring.
But it's just a possibility to reason about of course.

**Comment by njy** [comment]
> @njy The approach of looking at monikers does not address cases where you want to ship a stable package with one or more APIs still marked as preliminary. It also doesn't allow the granularity to be controlled (e.g. the use of a preliminary API would force the entire assembly to be preliminary). NuGet already provides direct validation for prerelease packages at this granularity, so I don't see the value in duplicating it elsewhere.

Yep I agree, as I said initially that was my point and the reason for this proposal.
The point now, methinks, is to determine if this is a scenario that we want to endorse and support or not because there are better ways to do it.
My observations based on personal experience was that people in fact do sometimes release stable packages with some experimental parts, but without the ability to clearly mark in a machine readable way parts of them as experimental.
Some seem to agree, some don't. It would be interesting to gater more opinions on this.

> The [BannedApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/tree/master/src/Microsoft.CodeAnalysis.BannedApiAnalyzers) package could be updated with new analyzers for this scenario. It already contains shared infrastructure for examining code to ensure disallowed APIs (for an extensible definition of disallowed) are not used.
> 
> It would not be difficult to expand the analysis to cover the rules for applying `[Preliminary]` to code, and update [PublicApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/tree/master/src/PublicApiAnalyzers) to clearly indicate that these APIs are not considered "shipped".

I didn't know about those analyzers, I'll take a look as soon as I can, thanks!

**Comment by am11** [comment]
Roslyn seems to support one `ExperimentalAttribute` from UWP's `Windows.Foundation.Metadata` namespace: https://github.com/dotnet/roslyn/blob/94ea302/src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs#L550

This doc highlights comparison with `Obsolete` and `Deprecated` attributes: https://github.com/dotnet/roslyn/blob/94ea302/docs/features/ExperimentalAttribute.md#obsoleteattribute-and-deprecatedattribute

Some related discussion in the original PR: https://github.com/dotnet/roslyn/pull/18708.
